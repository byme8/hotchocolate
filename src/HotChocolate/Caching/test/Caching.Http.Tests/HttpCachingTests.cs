using System;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Threading.Tasks;
using HotChocolate.Resolvers;
using HotChocolate.Types;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.TestHost;
using Microsoft.Extensions.DependencyInjection;
using Snapshooter.Xunit;
using Xunit;

namespace HotChocolate.Caching.Http.Tests;

public class HttpCachingTests : ServerTestBase
{
    public HttpCachingTests(TestServerFactory serverFactory)
            : base(serverFactory)
    {
    }

    [Fact]
    public async Task Test()
    {
        TestServer server = CreateServer(services =>
        {
            services.AddGraphQLServer()
                .UseQueryResultCachePipeline()
                .AddHttpQueryCache()
                .AddQueryType<Query>();
        });

        HttpClient client = server.CreateClient();

        GraphQLResult result = await client.PostQueryAsync("{ field1 }");

        result.MatchSnapshot();
    }

    private class Query
    {
        [CacheControl(2000)]
        public string Field1() => "Test";
    }
}

public class GraphQLResult
{
    public HttpResponseHeaders Headers { get; set; }

    public HttpContentHeaders ContentHeaders { get; set; }

    public string Body { get; set; }
}

internal static class TestServerExtensions
{
    public static async Task<GraphQLResult> PostQueryAsync(this HttpClient client, string query)
    {
        var payload = $"{{ \"query\": \"{query}\" }}";

        var content = new StringContent(payload, Encoding.UTF8, "application/json");

        HttpResponseMessage response = await client.PostAsync("/graphql", content);

        var result = new GraphQLResult
        {
            Headers = response.Headers,
            ContentHeaders = response.Content.Headers,
            Body = await response.Content.ReadAsStringAsync()
        };

        return result;
    }
}