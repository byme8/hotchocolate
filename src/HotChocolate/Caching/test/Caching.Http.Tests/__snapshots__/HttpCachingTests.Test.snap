{
  "Headers": [
    {
      "Key": "Cache-Control",
      "Value": [
        "public, max-age=2000"
      ]
    }
  ],
  "ContentHeaders": [
    {
      "Key": "Content-Type",
      "Value": [
        "application/json; charset=utf-8"
      ]
    }
  ],
  "Body": "{\"data\":{\"field1\":\"Test\"}}"
}
